compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/QueryTypeChecker.java

```diff
@@ -291,82 +294,148 @@ public BType resolveQueryType(SymbolEnv env, BLangExpression selectExp, BType ta
-                resolvedType = getResolvedType(queryResultType, type, isReadonly, env);
-                break;
-            case TypeTags.TABLE:
-                selectType = checkExpr(selectExp, env, types.getSafeType(((BTableType) type).constraint,
-                        true, true), data);
-                resolvedType = getResolvedType(symTable.tableType, type, isReadonly, env);
-                break;
-            case TypeTags.STREAM:
-                selectType = checkExpr(selectExp, env, types.getSafeType(((BStreamType) type).constraint,
-                        true, true), data);
-                resolvedType = symTable.streamType;
-                break;
-            case TypeTags.MAP:
-                List<BTupleMember> memberTypeList = new ArrayList<>(2);
-                BVarSymbol stringVarSymbol = new BVarSymbol(0, null, null,
-                        symTable.semanticError, null, symTable.builtinPos, SymbolOrigin.VIRTUAL);
-                memberTypeList.add(new BTupleMember(symTable.stringType, stringVarSymbol));
-                BType memberType = ((BMapType) type).getConstraint();
-                BVarSymbol varSymbol = Symbols.createVarSymbolForTupleMember(memberType);
-                memberTypeList.add(new BTupleMember(memberType, varSymbol));
-                BTupleType newExpType = new BTupleType(null, memberTypeList);
-                selectType = checkExpr(selectExp, env, newExpType, data);
-                resolvedType = getResolvedType(selectType, type, isReadonly, env);
-                break;
-            case TypeTags.STRING:
-            case TypeTags.XML:
-            case TypeTags.XML_COMMENT:
-            case TypeTags.XML_ELEMENT:
-            case TypeTags.XML_PI:
-            case TypeTags.XML_TEXT:
-                selectType = checkExpr(selectExp, env, type, data);
-                resolvedType = selectType;
-                break;
-            case TypeTags.INTERSECTION:
-                type = ((BIntersectionType) type).effectiveType;
-                solveSelectTypeAndResolveType(queryExpr, selectExp, type, collectionType, selectTypes,
-                        resolvedTypes, env, data, Symbols.isFlagOn(type.flags, Flags.READONLY));
-                return;
-            case TypeTags.NONE:
-            default:
-                // contextually expected type not given (i.e var).
-                selectType = checkExprSilent(nodeCloner.cloneNode(selectExp), type, data);
-                if (selectType != symTable.semanticError) {
-                    selectType = checkExpr(selectExp, env, type, data);
-                }  else {
-                    selectType = checkExpr(selectExp, env, data);
+        LinkedHashSet<BType> possibleSelectTypes = new LinkedHashSet<>();
+        List<BType> possibleResolvedTypes = new ArrayList<>();
+        LinkedHashSet<BType> errorTypes = new LinkedHashSet<>();
+        for (BType expType : expTypes) {
+            BType selectType, resolvedType;
```

Comment:
I think we can define these outside the loop since they're always assigned to before being used.